pyvrp/cpp/search/SwapStar.cpp

```diff
@@ -221,25 +120,28 @@ Cost SwapStar::evaluate(Route *routeU,
 {
     best = {};
 
-    for (auto *route : {routeU, routeV})
-        if (updated[route->idx()])
-        {
-            updateRemovalCosts(route, costEvaluator);
-            updated[route->idx()] = false;
+    if (updated[routeU->idx()])
+        updateRemovalCosts(routeU, costEvaluator);
 
-            for (size_t idx = data.numDepots(); idx != data.numLocations();
-                 ++idx)
-                cache(route->idx(), idx).shouldUpdate = true;
-        }
+    if (updated[routeV->idx()])
+        updateRemovalCosts(routeV, costEvaluator);
 
     for (auto *U : *routeU)
         for (auto *V : *routeV)
         {
+            // The following lines compute a delta cost of removing U and V from
+            // their own routes and inserting them into the other's route. Note
+            // that this is an approximation: the removal and insertions are
+            // evaluated in isolation, not taking into account that while U
+            // leaves its route, V will be inserted (and vice versa).
             Cost deltaCost = 0;
 
-            // Approximates the actual load changes - this is far from exact
-            // when there are also pickups. So it's pretty rough, but fast and
-            // seems to work well enough for most instances.
+            // Separating removal and insertion means particularly that the
+            // effects on load are counted double when inserting: U is still in
+            // the route, and now V is added as well. The following attempts to
+            // address this issue, but it is far from exact when there are both
+            // pickups and deliveries in the data. So it's pretty rough but fast
+            // and seems to work well enough for most instances.
             ProblemData::Client const &uClient = data.location(U->client());
             ProblemData::Client const &vClient = data.location(V->client());
```

Comment:
I figured out why the load stuff should not be part of the remove/insert evaluations, and documented that here. Annoyingly, this does mean we cannot completely decouple cost evaluations from `SwapStar`. But the impact is fairly limited so I can live with it.